### ISSUE TITLE
large model size and many zeros in cat_threshold field

### ISSUE THREAD

**Issue by JisongXie** [microsoft/LightGBM#4237]
## Description

I train a lightgbm classification model using lgb.train()，and save the model. My n_estimator is 3000, and it doesn't meet the early stopping and save the model in 3000th iteration. However, the model is oddly large，about 220M。

My dataset 's shape is as follow，
train data length is 847528. val data length is 211882.
train_x 's shape：(847528, 10)  train_y 's shape：(847528,)
val_x's shape：(211882, 10) val_y 's shape：(211882,)

When I show the model file， I see so many zeros in `cat_threshold` of each tree，which I haven't seen in other model files.
![image](https://user-images.githubusercontent.com/28192557/116383565-8b6a2380-a849-11eb-8876-0925d1dea8b1.png)

Anyone who knows why？

## Environment info

I use docker, and install miniconda3
```shell
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
/bin/bash Miniconda3-latest-Linux-x86_64.sh -f -b -p $CONDA_DIR
```

Command(s) you used to install LightGBM

```shell
git clone --recursive --branch stable --depth 1 https://github.com/Microsoft/LightGBM && \
cd LightGBM/python-package && python setup.py install && \
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM! Is it possible to provide a reproducible example (some simple code that does not rely on private data, which produces a model file like this)?

Also when you say "which I haven't seen in other model files" do you mean:

* The same code with an earlier version of LightGBM does not produce a model file like this
* Similar code with some other library (like `xgboost` or `catboost`) produces a different type of model file
* something else

**Comment by shiyu1994** [comment]
@JisongXie Thanks for using LightGBM.

First, let's make it clear what the `cat_threshold` entry means in the model file. In fact, it is a bit vector masking which values of the categorical feature used for split should be split into the left child node. Bit vectors for different internal nodes are separated by the `cat_boundaries`. For example, in your model file above, `cat_boudaries = 0 8754 14381 ...`. That means, the `[0, 8754)` items in `cat_threshold` belong to the internal node `0`, `[8743, 14381)` items in `cat_threshold` belong to the internal node `1`.

Each integer in `cat_threshold` is a 32-bit interger, and all the integers belonging to the same internal node make up the bit vector of length `32 * n`, where `n` is the number of items in `cat_threshold` belonging to that node. When a categorical feature value `k` should be split to the left child, the `k`th bit of the bit vector for that internal node is marked as `1`. All other bits are masked as `0`.

Your model has a lot of `0`s in the `cat_threshold` entry. That means, your category values of categorical features perhaps are not continuous, and can have very large range. For example, something like `0, 101, 1331, ...`. Remapping the categorical feature values into more compact continuous range e.g., `0,1,2, ...` may solve your problem.

In fact, we have a warning for such too sparse (or `too discontinuous`) categorical feature value distribution.
https://github.com/microsoft/LightGBM/blob/f97aa86ea586437536871b790cb9f51b190630b0/src/io/bin.cpp#L445-L448

**Comment by JisongXie** [comment]
@jameslamb @shiyu1994 
Thanks very much! As @shiyu1994 said, values of one of my categorical features are not continuous，and have very large range. I make ordinal encoding on it and it works. The model size is about 14M, and there aren't many zeros any more.
